### PR TITLE
Mainly refactoring datacore + identity system

### PR DESCRIPTION
**Pull request by EvilDragonfiend**
## About The Pull Request
Refactors/fixes some identity issues
* Datacore is now more reliably saved and trackable based on three info - name, age, gender
* You can check more personal identity for yourself (check memory) and someone else (by examining them)
* better memory viewing

## Why It's Good For The Game
Datacore is unreliably working because it only tracks data based on a given name - which means it can cause some conflicts when there're multiple people whose name is all the same to themselves. This is pretty bad for the nature of our game system.

With solving this issue, I've improved some RP-aspect systems relating to age and gender. Players will now have more chances to do RP with their age and gender.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

image

Roundstart chat notification

-------------------

image

when you check your memory

-------------------

image

when you're an antag

-------------------

image

when you absorbed an antag

------------

image
image

showing ID card

-------------

image

examine
if they're unknown(obscured), this line will not be shown.

-------------

image

console

image
image

let's change theirs

image

now card shows something different

-------------

image

sec hud is pretty fine. there's no conflict... actually it still can happen, but it means players messed up datacore.

-----

(no screenshot)

cryo doesn't remove a different individual's data whose name is the same of you.

--------

image

agent card forging is fine

</details>

## Changelog
:cl:
refactor: minor refactor in mob codes for real gender and real age system
add: real gender, real age in mobs info
add: your memory will tell your real name, real age, real gender
add: ghosts' gender will be based on mob's real gender.
add: data record system is more reliable. It will track an individual based on a name, an age, a gender, with these three info. This means you'll not get arrested by sec because of someone whose name is the same of you and that one is under arrest.
code: a new proc 'find_datacore_individual'. this is used for what's said above here, but when its parameter 'find_similar' is given as TRUE, it will return similar data rather than finding an exact individual with three matched information.
add: an agent card now has age/gender info for datacore reliability.
tweak: Eternal Youth symptom will turn your age back to your original age when it's gone.
del: cryo no longer deletes clone datas in genetics.
add: memory will have some better UI
add: memory will show your current identity. Changelings will recognise their sex and age as well.
add: examing someone will tell you their assumed age, and their gender.
add: people's voice now has age/gender info. It is currently used to track their record in datacore for AI to identify them for now.
add: HoP console can insert an ID card age and gender info. This is necessary for now.
add: Showing your ID card will tell people registered age and gender info.
code: ID card now has a standard setting proc. Please use `set_basic_info_on_spawn()` proc when you spawn a character with an ID card.
code: corrected some identity mistracking situations.
tweak: agent card has no longer reset option.
tweak: devil's age is 666. (beside it will be not used right now.)
code: get_face_name() proc is now get_face_info() proc. it will return more informations: name, age, gender. also, it's now located in carbon.dm
code: GetVoice_info proc has been added. People will recognise a sayer's age and gender from their voice (but only for datacore system. there's no visual examine way for that.)
/:cl:

### CODE REVIEW

**Comment by EvilDragonfiend** [comment]
Now it's ready to be reviewed but might need some testmerge. I tested cases at my side myself, but I am not certain for this one.

**Comment by EvilDragonfiend** [comment]
Leaves discord discussion

<details>
<summary>[Click here to see screenshot</summary>

Screenshot_20220919-094125_Discord

</details>

**Comment by itsmeow** [comment]
Really like this, the datacore and records systems are a complete mess and need a rework. This is a good first step.

**Comment by PowerfulBacon** [comment]
I agree that datacore is a bit of a mess but I only think this serves to worsen the problem. We are introducing 2 new variables that need to be handled on all mobs just because of data core, more work needs to be done to setup IDs and it's still not perfect

**Comment by EvilDragonfiend** [comment]
> Real_age and real_genders are my main issue with this since it adds a lot of work to a module

I think I've been adding too much to it unnecessarily. might have been enough with 'real_age = age' at init proc.

**Comment by EvilDragonfiend** [comment]
I suddenly remember why I was against ID based too.
Datacore can be removed and if you recover people's records after theirs are removed, you can't say if the new record has the original ID.

**Comment by EvilDragonfiend** [comment]
Going to remove unnecessary real var declaration for each mob

**Comment by EvilDragonfiend** [comment]
> I agree that datacore is a bit of a mess but I only think this serves to worsen the problem. 

It can't worsen the issue. Let me explain below.

> We are introducing 2 new variables that need to be handled on all mobs

It is true. I am going to remove unnecessary bloat codes. real_age = age stuff will be enough in general

> just because of data core, 

The main purpose of new variales is datacore, of course, but this can be initial support for more design that can handle your age/gender with more disguise option.
For example, what if we can have a human face mask from butched human skin?

> more work needs to be done to setup IDs

I like it though

> and it's still not perfect. 

Datacore isn't a perfect system and it would never be perfect. Like I stated above, IDs are even not reliable in datacore, can be removed or changed too.
Unreliable system should be handled more reliable way that is intuitive in game design.
This is why I am claming it can't worsen the issue.

> The only time I've seen 2 people with the same name is from clones and this wouldn't help that.

Although it is true that having the same name in a round would happen rare, but I don't think that can't be a reason not to have for better reliable datacore, like, wouldn't it be good for us to have more reliable system for more high pop? If we're populated enough, such issue would happen often.

> Real_age and real_genders are my main issue with this since it adds a lot of work to a module--

To be fixed

> -- that isn't datacore and provides a very minimal gain (Ive never seen two people with the same name except from whay i mentioned above and that would have its own set of issues)

Wouldn't that be a very minimal gain. Let me edit the codes first and let's see.

File: code/game/machinery/computer/medical.dm [hunk 1]
```diff
@@ -274,10 +274,14 @@
...
-							else
-								src.active1.fields["sex"] = "Male"
+							if(active1.fields["sex"] == MALE)
+								active1.fields["sex"] = FEMALE
+							if(active1.fields["sex"] == FEMALE)
+								active1.fields["sex"] = PLURAL
+							if(active1.fields["sex"] == PLURAL)
+								active1.fields["sex"] = NEUTER
+							if(active1.fields["sex"] == NEUTER)
+								active1.fields["sex"] = MALE
```
Comment (EvilDragonfiend) [inline code/game/machinery/computer/medical.dm:284]: Need to be else if

File: code/game/machinery/computer/security.dm [hunk 2]
```diff
@@ -767,10 +767,14 @@ What a mess.*/
...
-								active1.fields["sex"] = "Male"
+							if(active1.fields["sex"] == MALE)
+								active1.fields["sex"] = FEMALE
+							if(active1.fields["sex"] == FEMALE)
+								active1.fields["sex"] = PLURAL
+							if(active1.fields["sex"] == PLURAL)
+								active1.fields["sex"] = NEUTER
+							if(active1.fields["sex"] == NEUTER)
+								active1.fields["sex"] = MALE
 					if("age")
```
Comment (EvilDragonfiend) [inline code/game/machinery/computer/security.dm:778]: This too

File: code/__DEFINES/mobs.dm [hunk 3]
```diff
@@ -439,3 +439,6 @@
 #define THROW_MODE_DISABLED 0
 #define THROW_MODE_TOGGLE 1
 #define THROW_MODE_HOLD 2
+
+
+#define GENDER_OPTIONS list(MALE, FEMALE, PLURAL, NEUTER)
```
Comment (EvilDragonfiend) [inline code/__DEFINES/mobs.dm]: Maybe misc defines

File: code/datums/datacore.dm [hunk 4]
```diff
@@ -310,3 +311,53 @@
...
+			if(!A)
+				A = R // [single match]
+		if(age_gap) // being 0 means you have a best case for the age. no need to compare age.
+			var/age_abs = abs(age-R.fields["age"])
+			if(age_abs < age_gap)
+				age_gap = age_abs
+				A = R   // [half match], but...↓
+			if(age_gap) // [similar match], if age_gap exists.  (and, indentation is correct. don't change.)
+				continue
+		if(R.fields["sex"] == gender) // if it didn't get continue, [half match] will check gender integrity to be [full match]
```
Comment (EvilDragonfiend) [inline code/datums/datacore.dm]: Should have !age_gap

File: code/game/objects/items/cards_ids.dm [hunk 5]
```diff
@@ -109,11 +109,13 @@
 	slot_flags = ITEM_SLOT_ID
 	armor = list("melee" = 0, "bullet" = 0, "laser" = 0, "energy" = 0, "bomb" = 0, "bio" = 0, "rad" = 0, "fire" = 100, "acid" = 100, "stamina" = 0)
 	resistance_flags = FIRE_PROOF | ACID_PROOF
-	var/mining_points = 0 //For redeeming at mining equipment vendors
-	var/list/access = list()
 	var/registered_name// The name registered_name on the card
+	var/age = 13
```
Comment (EvilDragonfiend) [inline code/game/objects/items/cards_ids.dm]: Refistered age var name wojld be better

Comment (EvilDragonfiend) [reply code/game/objects/items/cards_ids.dm]: It shouldn't be 13 as default.

File: code/game/objects/items/cards_ids.dm [hunk 6]
```diff
@@ -434,7 +442,7 @@ update_label("John Doe", "Clowny")
 			if(user.mind.special_role || anyone)
 				to_chat(usr, "<span class='notice'>The card's microscanners activate as you pass it over the ID, copying its access.</span>")
 
-/obj/item/card/id/syndicate/attack_self(mob/user)
+/obj/item/card/id/syndicate/attack_self(mob/living/user)
```
Comment (EvilDragonfiend) [inline code/game/objects/items/cards_ids.dm:445]: Maybe, L = user?

File: code/game/objects/items/cards_ids.dm [hunk 7]
```diff
@@ -382,7 +389,8 @@ update_label("John Doe", "Clowny")
 	icon_state = "syndicate"
 	hud_state = JOB_HUD_SYNDICATE
 	var/anyone = FALSE //Can anyone forge the ID or just syndicate?
-	var/forged = FALSE //have we set a custom name and job assignment, or will we use what we're given when we chameleon change?
```
Comment (EvilDragonfiend) [inline code/game/objects/items/cards_ids.dm:385]: Not docu

File: code/modules/antagonists/changeling/powers/absorb.dm [hunk 8]
```diff
@@ -66,7 +66,7 @@
 		owner.copy_languages(target, LANGUAGE_ABSORB)
 
 		var/datum/mind/suckedbrain = target.mind
-		user.mind.memory += "<BR><b>We've absorbed [target]'s memories into our own...</b><BR>[suckedbrain.memory]<BR>"
+		user.mind.memory += "<b>We've absorbed [target?.mind.name ? target.mind.name : target.real_name]'s memories into our own...</b><BR>[suckedbrain.memory]<BR>"
```
Comment (EvilDragonfiend) [inline code/modules/antagonists/changeling/powers/absorb.dm]: Need to be better

File: code/modules/antagonists/devil/true_devil/_true_devil.dm [hunk 9]
```diff
@@ -9,6 +9,7 @@
 	icon = 'icons/mob/32x64.dmi'
 	icon_state = "true_devil"
 	gender = NEUTER
+	age = 666
```
Comment (EvilDragonfiend) [inline code/modules/antagonists/devil/true_devil/_true_devil.dm:12]: Undocu

File: code/modules/mob/living/carbon/alien/alien.dm [hunk 10]
```diff
@@ -2,6 +2,7 @@
 	name = "alien"
 	icon = 'icons/mob/alien.dmi'
 	gender = FEMALE //All xenos are girls!!
+	age = 0
```
Comment (EvilDragonfiend) [inline code/modules/mob/living/carbon/alien/alien.dm]: To be removed? Or keep it?

File: code/modules/mob/living/carbon/carbon.dm [hunk 11]
```diff
@@ -43,6 +43,22 @@
 		if(H)
 			H.update_icon()
 
+//Returns "Unknown" if facially disfigured and real_name if not. Useful for setting name when Fluacided or when updating a human's name variable
+/mob/living/carbon/proc/get_face_info(if_no_face="Unknown", target=RETURNS_NAME)
```
Comment (EvilDragonfiend) [inline code/modules/mob/living/carbon/carbon.dm:47]: Undocumented proc relocation

**Comment by EvilDragonfiend** [comment]
Need to check clone

**Comment by EvilDragonfiend** [comment]
It's hard to post the testing evidence for voice age/gender tracking and obscured age/gender tracking, but it's tested, and the test result is very nice.

**Comment by EvilDragonfiend** [comment]
ah fuck

**Comment by EvilDragonfiend** [comment]
I don't want to be bothered with this for now... maybe opened later.

**Pull request closed without merging**